primitives/src/merkle_tree/internal.rs

```diff
@@ -0,0 +1,672 @@
+use typenum::Unsigned;
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize, Deserialize)]
+pub enum MerkleNode<E: Element, I: Index, T: NodeValue> {
+    Empty,
+    Branch {
+        value: T,
+        children: Vec<Box<MerkleNode<E, I, T>>>,
+    },
+    Leaf {
+        value: T,
+        pos: I,
+        elem: E,
+    },
+    ForgettenSubtree {
+        value: T,
+    },
+}
+
+impl<E, I, T> MerkleNode<E, I, T>
+where
+    E: Element,
+    I: Index,
+    T: NodeValue,
+{
+    /// Return the value of this [`MerkleNode`].
+    #[inline]
+    pub(crate) fn value(&self) -> T {
+        match self {
+            Self::Empty => T::default(),
+            Self::Leaf {
+                value,
+                pos: _,
+                elem: _,
+            } => *value,
+            Self::Branch { value, children: _ } => *value,
+            Self::ForgettenSubtree { value } => *value,
+        }
+    }
+}
+
+/// A merkle commitment consists a root hash value, a tree height and number of
+/// leaves
+#[derive(
+    Eq, PartialEq, Clone, Copy, Ord, PartialOrd, Hash, CanonicalSerialize, CanonicalDeserialize,
+)]
+#[tagged_blob("MERKLE_COMM")]
+pub struct MerkleTreeCommitment<T: NodeValue> {
+    /// Root of a tree
+    digest: T,
+    /// Height of a tree
+    height: usize,
+    /// Number of leaves in the tree
+    num_leaves: u64,
+}
+
+impl<T: NodeValue> MerkleTreeCommitment<T> {
+    pub fn new(digest: T, height: usize, num_leaves: u64) -> Self {
+        MerkleTreeCommitment {
+            digest,
+            height,
+            num_leaves,
+        }
+    }
+}
+
+impl<T: NodeValue> MerkleCommitment<T> for MerkleTreeCommitment<T> {
+    fn digest(&self) -> T {
+        self.digest
+    }
+
+    fn height(&self) -> usize {
+        self.height
+    }
+
+    fn size(&self) -> u64 {
+        self.num_leaves
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize, Deserialize)]
+pub struct MerkleProof<E, I, T, Arity>
+where
+    E: Element,
+    I: Index,
+    T: NodeValue,
+    Arity: Unsigned,
+{
+    /// Proof of inclusion for element at index `pos`
+    pub pos: I,
+    /// Nodes of proof path, from root to leaf
+    pub proof: Vec<MerkleNode<E, I, T>>,
+
+    /// Place holder for Arity
+    _phantom_arity: PhantomData<Arity>,
+}
+
+impl<E, I, T, Arity> MerkleProof<E, I, T, Arity>
+where
+    E: Element,
+    I: Index,
+    T: NodeValue,
+    Arity: Unsigned,
+{
+    pub fn tree_height(&self) -> usize {
+        self.proof.len()
+    }
+
+    pub fn new(pos: I, proof: Vec<MerkleNode<E, I, T>>) -> Self {
+        MerkleProof {
+            pos,
+            proof,
+            _phantom_arity: PhantomData,
+        }
+    }
+}
+
+#[allow(clippy::type_complexity)]
+pub(crate) fn build_tree_internal<E, H, I, Arity, T>(
+    height: usize,
+    elems: impl IntoIterator<Item = impl Borrow<E>>,
+) -> Result<(Box<MerkleNode<E, I, T>>, u64), PrimitivesError>
+where
+    E: Element,
+    H: DigestAlgorithm<E, I, T>,
+    I: Index + From<u64>,
```

Comment:
There's a lot of `Index + From<u64>`, maybe we should add to `trait Index: ... + From<u64>` instead?

Reply:
@alxiong @mrain this prevents using a complex type as `Index` (e.g. for key-value Merkle tree) are we ok with that?

Reply:
I agree with @jbearer . It's not changed.